File: python/sorting/bubblesort.py
#!/usr/bin/env python3

"""
Bubblesort
Version: 1
Desc: Goes through the list (multiple times) and compares 
      adjacent items and swaps those that are out of order.
"""

def bubble(list):

    n = len(list)
    swapped = True
    while swapped:
        swapped = False
        for i in range(1, n):
            if list[i-1] > list[i]:
                # Swap
                temp_value = list[i-1]
                list[i-1] = list[i]
                list[i] = temp_value
                swapped = True
        n = n-1

    return list
    

#print(bubble([1, 2, 3, 4, 5, 6, 7]))
print(bubble([5, 3, -1, 13, 6, 2, 8]))

File: java/sorting/BubbleSort.java
class BubbleSort {

    public static <E extends Comparable> void sort(E[] array) {
        int upper = array.length-1;
        boolean swapped;
        
        do {
            swapped = false;
            for (int i = 0; i < upper; i++) {
                if (array[i].compareTo(array[i+1]) > 0) {
                    Swap.index(array, i, i+1);
                    swapped = true;
                }
            }
            upper--;
        } while (swapped);
    }
}

/* Alternative version without while loop
class BubbleSort {
    private static int[] sort(int[] array) {
        int temp_value;
        boolean swapped = true;

        // For-loop instead of while
        for (int i = array.length - 1; swapped && i >= 0; i--) {
            swapped = false;
            for (int j = 0; j < i; j++) {
                if (array[j] > array[j+1]) { // Out of order, swap
                    temp_value = array[j];
                    array[j] = array[j+1];
                    array[j+1] = temp_value;
                    swapped = true;
                }
            }
        }

        return array;
    }
}
*/
